be/src/exprs/ngram.cpp

```diff
@@ -0,0 +1,293 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// This file is based on code available under the Apache license here:
+//  https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/FunctionsStringSimilarity.cpp
+
+#include "column/column_hash.h"
+#include "exprs/string_functions.h"
+namespace starrocks {
+static constexpr size_t max_string_size = 1 << 15;
+// uint16[2^16] can almost fit into L2
+static constexpr size_t map_size = 1 << 16;
+// we restrict needle's size smaller than 2^16, so even if every gram in needle is the same as each other
+// we still only need one uint16 to store its frequency
+using NgramHash = uint16;
+
+struct Ngramstate {
+    using DriverMap = phmap::parallel_flat_hash_map<int32_t, std::unique_ptr<NgramHash[]>, phmap::Hash<int32_t>,
+                                                    phmap::EqualTo<int32_t>, phmap::Allocator<int32_t>,
+                                                    NUM_LOCK_SHARD_LOG, std::mutex>;
+    std::unique_ptr<NgramHash[]> publicHashMap;
+    DriverMap driver_maps; // hashMap for each pipeline_driver, to make it driver-local
+
+    size_t needle_gram_count = 0;
+
+    NgramHash* get_or_create_driver_hashmap() {
+        int32_t driver_id = CurrentThread::current().get_driver_id();
+        NgramHash* res = nullptr;
+        if (LIKELY(driver_maps.contains(driver_id))) {
+            res = driver_maps[driver_id].get();
+        } else {
+            auto per_driver_map = std::make_unique<NgramHash[]>(map_size);
+            std::memcpy(per_driver_map.get(), publicHashMap.get(), map_size * sizeof(NgramHash));
+            res = per_driver_map.get();
+            driver_maps.lazy_emplace(driver_id, [&](auto build) { build(driver_id, std::move(per_driver_map)); });
+        }
+
+        DCHECK(!!res);
+        return res;
+    }
+};
+
+template <size_t N, bool case_insensitive, bool use_utf_8, class Gram>
+class NgramFunctionImpl {
+public:
+    StatusOr<ColumnPtr> static ngram_search_impl(FunctionContext* context, const Columns& columns) {
+        RETURN_IF_COLUMNS_ONLY_NULL(columns);
+        const ColumnPtr& haystack_ptr = columns[0];
```

Comment:
1. use auto as possible as you can:
2. use xxx_column instead of xxx_ptr
```cpp
const auto & haystack_column = columns[0];
const auto & needle_column = columns[1];
```